### PR TITLE
feat: add MegaChunkGDN fused operator for Qwen3.5 on npu device.

### PR DESCRIPTION
**Pull request by Sinle4Cat**
## Description

Adds the NPU MegaChunkGDN fused operator path for Qwen3.5 Gated Delta Net.

This PR wires the new NPU operator through xLLM kernel APIs, metadata, and the Qwen3.5 GDN layer path to reduce overhead in the gated delta net execution path. It also removes an unrelated global `-Wno-error=maybe-uninitialized` CMake change after verifying the NPU build passes without it.

## Related Issues

None.

## Change Type

- [ ] Bug fix
- [x] New feature
- [x] Performance improvement
- [ ] Refactor
- [ ] Documentation
- [ ] Test
- [ ] Build or CI

## Pull Request Checklist

Thank you for contributing to xLLM. Before requesting review, please make sure the following items are complete.

### PR Title and Commit Messages

- [x] The PR title and each commit message follow the xLLM commit format: `<type>: <subject>`.

> Allowed types: feat, bugfix, docs, test, refactor, chore, style, revert, perf, model, build, release.  
> The subject should use clear English, start with a verb, include at least 4 words, and end with `.`.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` or an equivalent command.
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

### Self Review

- [ ] I have self-reviewed the code according to `.agents/skills/code-review/references/custom-code-style.md`, especially code written or assisted by AI.
- [ ] I have rebased this PR onto the latest `main` branch.

### Build and Test Coverage

- [ ] Tests have been added or updated as needed.
- [ ] CUDA: `python setup.py build test` has passed on a CUDA machine.
- [x] NPU: `python setup.py build test` has passed on an NPU machine.
- [ ] MLU: `python setup.py build test` has passed on an MLU machine.

## Reviewer Notes

Validation performed:
- `python setup.py build test` passed on an NPU machine.
- Qwen3.5-27B TP2 no-MTP、MTP 2k/2k concurrency=1 , 2 ,4 benchmark passed 4/4 requests on continuous NPU devices 
- Qwen3.5-27B TP2 no-MTP ceval dataset accuracy 92.59, MTP have a problem with 8 concurrent requests crashing unexpectedly

### CODE REVIEW

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <glog/logging.h>
+
+#include <unordered_map>
```
[the remark this answers is not in the document]
Comment (Sinle4Cat) [reply xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp:20]: have include mutex

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  auto out = torch::empty({B, T, H, V}, opts_fp16);
+  auto g_sum = torch::empty({B, T, H}, opts_fp32);
+  auto g_t = torch::empty({H, T}, opts_fp32);
+  auto beta_t = torch::empty({H, T}, opts_fp16);
+  auto a = torch::zeros({B, T, H, kMegaChunkSize}, opts_fp16);
+  auto a_inv_f32 = torch::zeros({B, T, H, kMegaChunkSize}, opts_fp32);
+  auto a_inv = torch::zeros({B, T, H, kMegaChunkSize}, opts_fp16);
+  auto w = torch::empty({B, T, H, V}, opts_fp16);
+  auto u = torch::empty({B, T, H, V}, opts_fp16);
+  auto h = torch::zeros({num_matrices, K, V}, opts_fp16);
```
Comment (zhang-minchao) [inline xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp:140]: 新算子在 Qwen3.5 prefill 路径上替换了 `chunk_gated_delta_rule`,两者对 state 的精度处理不同,建议补一份数值对比。

**可验证的差异**:被替换路径跨 chunk 的 state 进位器与 initial/final-state 的 I/O 是 **fp32**——见 tilelang kernel `chunk_gated_delta_rule_fwd_h.py`(`accum_dtype=float32`):进位累积器 `h_state_ub_float` 在 fp32 上做 decay(`*= exp(g)`)与 `+= k@v_new`,`h0`/`ht` 也是 fp32。而本算子的 `h`、`initial_state_arg`、`final_state` 都按 **fp16** 分配。

需要说明的是,per-chunk 的 matmul 两边都是低精度、基本对等:旧 kernel 里 state 进 `w@h` 前会先从 fp32 降成 bf16,per-chunk 的 `h_out` 本身也是 bf16(只是 MAC 在 fp32 L0C 上累加);新算子对应位置是 fp16。所以差异不在 matmul,而在**跨 chunk 的 state 进位与 initial/final-state I/O 从 fp32 变成了 fp16**。

**不确定点**:`aclnnMegaChunkGdn` 内部跨 chunk 累积用的精度,从这份 PR 无法确认。若内部仍用 fp32 累加、只是把 I/O 写成 fp16,则影响很小;若直接在 fp16 上累积,长序列(2k prefill ≈ 32 个 chunk)下可能漂移。PR notes 里的 “benchmark passed 4/4 requests” 只能说明能跑、能出 token,不能证明与被替换路径一致。

能否确认下 `aclnnMegaChunkGdn` 的内部累积精度?如果是 fp16 累积,建议补一份与 `chunk_gated_delta_rule` 的数值对比(逐 token logits / 输出 diff,或 PPL)来界定影响。另外 `param.h` 把 q/k/v 标注为 bfloat16,但 kernel 实际以 fp16 运行,请让注释与实现保持一致。

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    torch::Tensor& k,
+    torch::Tensor& v,
+    torch::Tensor& g,
+    torch::Tensor& beta,
+    const std::optional<float>& scale,
+    const std::optional<torch::Tensor>& initial_state,
+    bool output_final_state,
+    const std::optional<torch::Tensor>& cu_seqlens,
+    bool use_qk_l2norm_in_kernel) {
+  const auto input_dtype = q.scalar_type();
```
Comment (zhang-minchao) [inline xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp]: `scalar_type()` 返回 `torch::ScalarType`,是一个普通枚举(简单/基础类型)。按项目风格指南(第 3 节,“简单/基础类型不要用 `auto`”),并与本路径替换的算子保持一致——`ops_api.cpp:1639` 写的是 `const torch::ScalarType input_dtype = params.q.scalar_type();`——这里应显式写出类型:
```cpp
const torch::ScalarType input_dtype = q.scalar_type();
```

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  const auto k_dtype = k.scalar_type();
+  const auto v_dtype = v.scalar_type();
```
Comment (zhang-minchao) [inline xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp]: `k_dtype`(第 76 行)和 `v_dtype`(第 77 行)声明后从未使用,第 125 行的 `const int64_t Hqk = k.size(2);` 同样如此。未使用的局部变量会触发 `-Wunused-variable`,在 `-Werror` 配置下会导致构建失败。请删除这三个。

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  // Apply L2 normalization to q and k if requested
+  torch::Tensor q_normalized = q;
+  torch::Tensor k_normalized = k;
+  if (use_qk_l2norm_in_kernel) {
+    auto q_norm =
+        torch::linalg_norm(q, /*ord=*/2, /*dim=*/-1, /*keepdim=*/true);
+    q_normalized = q / q_norm.clamp_min(1e-6);
+    auto k_norm =
+        torch::linalg_norm(k, /*ord=*/2, /*dim=*/-1, /*keepdim=*/true);
+    k_normalized = k / k_norm.clamp_min(1e-6);
```
Comment (zhang-minchao) [inline xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp]: 这里的 L2norm 是在 GDN kernel 之外用 eager torch 算子完成的(`linalg_norm` + `clamp_min` + 除法),每次调用是好几次 NPU 算子 launch。本路径替换的算子在 `use_qk_l2norm_in_kernel` 为 true 时走的是融合的单次 Triton kernel `npu::npu_l2norm_last_dim`(`ops_api.cpp:1662-1667`;实现见 `third_party/torch_npu_ops/triton_npu/torch_api/npu_triton_l2norm.cpp`,内部一次 kernel launch 完成平方和 + eps + 除法)。鉴于本 PR 的目标是降低开销,改用 `npu_l2norm_last_dim` 可以省掉这些额外的 launch;它内部通过 view 支持任意前置维度,q/k 的 `[B,T,H,K]` 可以直接传入。

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp (hunk 2 above)
Comment (Sinle4Cat) [reply xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp:140]: In xllm_ops implementation, the cross-chunk recurrent update is not pure fp16 accumulation. In chunk_h.cpp, the running state s_ub, kv_ub, gate values, and decay coefficients
  are float UB tiles. The recurrence S = exp(g_last) * S + KV is computed with float vector ops, and Cube GEMMs use float accumulators.

  There is still fp16 storage at the GM/workspace boundary: state snapshots, final_state, and several intermediate workspaces are stored as fp16 buffers, so there is fp16 rounding when spilling. To
  bound this effect, I added operator-level comparison against chunk_gated_delta_rule, including 2k prefill. The 2k cases pass allclose; output max abs is 1.2207e-4 and final_state max abs is <=
  5.65119e-4.

Comment (Sinle4Cat) [reply xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp:140]: | Case | Tokens | Initial state | Output max abs | Output mean abs | Output allclose | State max abs | State mean abs | State allclose |
| --- | ---: | --- | ---: | ---: | --- | ---: | ---: | --- |
| qwen35_tp2_short_zero_state | 64 | zero | 6.10352e-05 | 3.95746e-06 | true | 0.000538894 | 6.01922e-05 | true |
| qwen35_tp2_two_chunk_zero_state | 128 | zero | 6.10352e-05 | 5.41153e-06 | true | 0.000597477 | 6.78504e-05 | true |
| qwen35_tp2_multi_chunk_zero_state | 257 | zero | 6.10352e-05 | 6.29372e-06 | true | 0.000532078 | 7.0497e-05 | true |
| qwen35_tp2_varlen_zero_state | 450 | zero | 6.10352e-05 | 5.5636e-06 | true | 0.000558423 | 6.51165e-05 | true |
| qwen35_tp2_2k_zero_state | 2048 | zero | 0.00012207 | 7.07793e-06 | true | 0.000565119 | 6.80714e-05 | true |
| qwen35_tp2_short_nonzero_state | 64 | nonzero | 6.10352e-05 | 4.55601e-06 | true | 0.000518747 | 5.93633e-05 | true |
| qwen35_tp2_varlen_nonzero_state | 450 | nonzero | 6.10352e-05 | 5.95965e-06 | true | 0.000614885 | 6.4692e-05 | true |
| qwen35_tp2_2k_nonzero_state | 2048 | nonzero | 0.00012207 | 7.15477e-06 | true | 0.00050658 | 6.97534e-05 | true |

File: xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp (hunk 4 above)
Comment (Sinle4Cat) [reply xllm/core/kernels/npu/xllm_ops/npu_mega_chunk_gdn.cpp]: 已修复

**Review by zhang-minchao — APPROVED** [review]

**Review by yingxudeng — APPROVED** [review]
